File: auction-house/js/src/instructions/updateAuctionHouse.ts
import { AccountMeta, PublicKey, TransactionInstruction } from '@solana/web3.js';
import * as beet from '@metaplex-foundation/beet';

export type UpdateAuctionHouseInstructionArgs = {
  sellerFeeBasisPoints: beet.COption<number>;
  requiresSignOff: beet.COption<boolean>;
  canChangeSalePrice: beet.COption<boolean>;
};
const updateAuctionHouseInstructionArgsStruct =
  new beet.BeetArgsStruct<UpdateAuctionHouseInstructionArgs>([
    ['sellerFeeBasisPoints', beet.coption(beet.u16)],
    ['requiresSignOff', beet.coption(beet.bool)],
    ['canChangeSalePrice', beet.coption(beet.bool)],
  ]);
export type UpdateAuctionHouseInstructionAccounts = {
  treasuryMint: PublicKey;
  payer: PublicKey;
  authority: PublicKey;
  newAuthority: PublicKey;
  feeWithdrawalDestination: PublicKey;
  treasuryWithdrawalDestination: PublicKey;
  treasuryWithdrawalDestinationOwner: PublicKey;
  auctionHouse: PublicKey;
  tokenProgram: PublicKey;
  systemProgram: PublicKey;
  ataProgram: PublicKey;
  rent: PublicKey;
};

export function createUpdateAuctionHouseInstruction(
  accounts: UpdateAuctionHouseInstructionAccounts,
  args: UpdateAuctionHouseInstructionArgs,
) {
  const {
    treasuryMint,
    payer,
    authority,
    newAuthority,
    feeWithdrawalDestination,
    treasuryWithdrawalDestination,
    treasuryWithdrawalDestinationOwner,
    auctionHouse,
    tokenProgram,
    systemProgram,
    ataProgram,
    rent,
  } = accounts;

  const [data, _] = updateAuctionHouseInstructionArgsStruct.serialize(args);
  const keys: AccountMeta[] = [
    {
      pubkey: treasuryMint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: payer,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: authority,
      isWritable: false,
      isSigner: true,
    },
    {
      pubkey: newAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: feeWithdrawalDestination,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: treasuryWithdrawalDestination,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: treasuryWithdrawalDestinationOwner,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: auctionHouse,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: tokenProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: systemProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: ataProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: rent,
      isWritable: false,
      isSigner: false,
    },
  ];

  const ix = new TransactionInstruction({
    programId: new PublicKey('hausS13jsjafwWwGqZTUQRmWyvyxn9EQpqMwV1PBBmk'),
    keys,
    data,
  });
  return ix;
}

File: auction-house/program/tests/update_auction_house.rs
#![cfg(test)]
pub mod utils;

use anchor_client::{
    solana_sdk::{signature::Keypair, signer::Signer, system_program, sysvar},
    ClientError,
};
use mpl_auction_house::{
    accounts as mpl_auction_house_accounts, instruction as mpl_auction_house_instruction,
    AuctionHouse,
};
use rand::rngs::OsRng;
use utils::setup_functions::{setup_auction_house, setup_client, setup_payer_wallet};

mod update_auction_house {

    use super::*;

    #[test]
    fn success() -> Result<(), ClientError> {
        // Client
        let client = setup_client(setup_payer_wallet());

        // Program
        let program = client.program(mpl_auction_house::id());

        // Auction house authority
        let authority_keypair = Keypair::generate(&mut OsRng);
        let authority = authority_keypair.pubkey();
        // New auction house authority
        let new_authority = Keypair::generate(&mut OsRng).pubkey();

        // Treasury mint key
        let t_mint_key = spl_token::native_mint::id();

        let auction_house_key = setup_auction_house(&program, &authority, &t_mint_key).unwrap();

        let twd_key = program.payer();
        let fwd_key = program.payer();
        let tdw_ata = twd_key;

        let seller_fee_basis_points: u16 = 345;

        program
            .request()
            .signer(&authority_keypair)
            .accounts(mpl_auction_house_accounts::UpdateAuctionHouse {
                treasury_mint: t_mint_key,
                payer: program.payer(),
                authority,
                new_authority,
                fee_withdrawal_destination: fwd_key,
                treasury_withdrawal_destination: tdw_ata,
                treasury_withdrawal_destination_owner: twd_key,
                auction_house: auction_house_key,
                token_program: spl_token::id(),
                system_program: system_program::id(),
                ata_program: spl_associated_token_account::id(),
                rent: sysvar::rent::id(),
            })
            .args(mpl_auction_house_instruction::UpdateAuctionHouse {
                seller_fee_basis_points: Some(seller_fee_basis_points),
                requires_sign_off: Some(false),
                can_change_sale_price: Some(false),
            })
            .send()?;

        let auction_house_account: AuctionHouse = program.account(auction_house_key)?;

        assert_eq!(
            seller_fee_basis_points,
            auction_house_account.seller_fee_basis_points
        );

        assert_eq!(false, auction_house_account.requires_sign_off);
        assert_eq!(false, auction_house_account.can_change_sale_price);

        Ok(())
    }
}
